tensorflow_probability/python/util/deferred_tensor_test.py

```diff
@@ -335,6 +335,17 @@ def testOperatorIter(self):
         for _ in iter(x):
           pass
 
+  def testMethodNumpy(self):
+    x_ = np.array([0., 1.])
+    x = tfp.util.DeferredTensor(tf.Variable(x_), tf.math.exp)
+
+    self.evaluate([v.initializer for v in x.trainable_variables])
+    if tf.executing_eagerly():
+      self.assertAllEqual(tf.math.exp(x_).numpy(), x.numpy())
+    else:
+      with self.assertRaises(AttributeError):
```

Comment:
This one raises:
```
Traceback (most recent call last):
  File "/Users/hartikainen/conda/envs/tensorflow-dev/lib/python3.7/site-packages/absl/testing/parameterized.py", line 263, in bound_param_test
    test_method(self, **testcase_params)
  File "/private/var/tmp/_bazel_hartikainen/4f63de4300ef6f519aa77402a0c94964/sandbox/darwin-sandbox/483/execroot/tensorflow_probability/bazel-out/darwin-fastbuild/bin/tensorflow_probability/python/util/deferred_tensor_test.runfiles/tensorflow_probability/tensorflow_probability/python/internal/test_combinations.py", line 312, in decorated
    execute_test_method()
  File "/private/var/tmp/_bazel_hartikainen/4f63de4300ef6f519aa77402a0c94964/sandbox/darwin-sandbox/483/execroot/tensorflow_probability/bazel-out/darwin-fastbuild/bin/tensorflow_probability/python/util/deferred_tensor_test.runfiles/tensorflow_probability/tensorflow_probability/python/internal/test_combinations.py", line 295, in execute_test_method
    test_method(**kwargs_to_pass)
  File "/private/var/tmp/_bazel_hartikainen/4f63de4300ef6f519aa77402a0c94964/sandbox/darwin-sandbox/483/execroot/tensorflow_probability/bazel-out/darwin-fastbuild/bin/tensorflow_probability/python/util/deferred_tensor_test.runfiles/tensorflow_probability/tensorflow_probability/python/util/deferred_tensor_test.py", line 352, in testMethodNumpy
    x.numpy()
  File "/private/var/tmp/_bazel_hartikainen/4f63de4300ef6f519aa77402a0c94964/sandbox/darwin-sandbox/483/execroot/tensorflow_probability/bazel-out/darwin-fastbuild/bin/tensorflow_probability/python/util/deferred_tensor_test.runfiles/tensorflow_probability/tensorflow_probability/python/util/deferred_tensor.py", line 293, in numpy
    return self._value().numpy()
AttributeError: 'Tensor' object has no attribute 'numpy'
```

Reply:
Hm, not sure why it's getting a `Tensor` and not an `EagerTensor` (only the latter has a `.numpy()`)

Reply:
Isn't that expected because it's running in graph mode?

Reply:
oh i was not looking at the x.numpy() below. i thought it was firing on the one inside the `if tf.executing_eagerly()` conditional. so is there a problem, or is it all good?

Reply:
No, it looks good to me. Just wanted to make sure that `AttributeError` is what we'd expect here.

Reply:
Ohh I see. We could  reasonably catch this and re-raise it with a more localized error message (eg, TypeError(?): .numpy() only works in eager mode)

Reply:
specifically,

```python
v = self._value()
if not hasattr(v, 'numpy'):
  raise TypeError("DeferredTensor.numpy() is only supported in eager execution mode")
return v.numpy()
```

This, rather than catching the exception, since it's best not to catch an re-raise (which can lose useful context if we accidentally catch something other than what we expected).

Reply:
Sounds good! It now raises `TypeError`.

Reply:
> specifically

Oh oops, I changed it before saw your comment. I changed it to:
```python
    try:
      return self._value().numpy()
    except AttributeError:
      if isinstance(self._value(), tf.Tensor):
        raise TypeError(
            "'{}.numpy()' only works in eager mode.".format(type(self).__name__))
      else:
        raise
```

Let me know if you prefer your version and I can change it.

Reply:
Thanks, yeah, I prefer mine  as it's a bit simpler. I do appreciate your effort to catch the most specific error and re-raise anything else!

Reply:
Cool, should be changed now. Thanks for the comments!